### PR TITLE
Lint attributes on function arguments

### LINKED ISSUE #61238: Lint attributes on function args have no effect
In order for the lint attributes to actually have an effect, we need to change `ast::visit` and `hir::intravisit` significantly before even being able to adjust `LateLintPass` and `EarlyLintPass` to handle the attributes.

The current schemes for iterating over the pattern of the argument and the type of the argument vary significantly between functions, methods extern functions, closures, .... I'm not sure if they can be unified easily, since extern functions don't allow patterns as arguments, but I think there should be a `visit_argument` method which all of the sites should call. In case attributes aren't allowed in extern functions, this should be possible, otherwise we'll have to figure out something.

After that, the `*LintPass` traits need a new `check_argument` method, too and before invocing said method, a `with_lint_attrs` call needs to be nested around the `check_argument` and `walk_argument` method calls.

_Originally posted by @oli-obk in https://github.com/rust-lang/rust/pull/60669#issuecomment-496031277_

### PR DESCRIPTION
**Pull request by c410-f3r**
Fixes #61238.

cc #60406

### CODE REVIEW

File: src/test/ui/suggestions/as-ref.stderr [hunk 1]
```diff
@@ -2,9 +2,10 @@ error[E0308]: mismatched types
   --> $DIR/as-ref.rs:6:27
    |
 LL |   opt.map(|arg| takes_ref(arg));
-   |       ---                 ^^^ expected &Foo, found struct `Foo`
-   |       |
-   |       help: consider using `as_ref` instead: `as_ref().map`
```
Comment (oli-obk) [inline src/test/ui/suggestions/as-ref.stderr]: This is a regression that should be addressed before the PR can be merged. Any clue what lead to it?

File: src/test/ui/mismatched_types/issue-38371.stderr [hunk 2]
```diff
@@ -6,7 +6,6 @@ LL | fn foo(&foo: Foo) {
    |
    = note: expected type `Foo`
               found type `&_`
-   = help: did you mean `foo: &Foo`?
```
Comment (oli-obk) [inline src/test/ui/mismatched_types/issue-38371.stderr]: Another regression, this time in type checking

File: src/libsyntax/visit.rs [hunk 3]
```diff
@@ -549,12 +550,8 @@ pub fn walk_fn_ret_ty<'a, V: Visitor<'a>>(visitor: &mut V, ret_ty: &'a FunctionR
 }
 
 pub fn walk_fn_decl<'a, V: Visitor<'a>>(visitor: &mut V, function_declaration: &'a FnDecl) {
-    for argument in &function_declaration.inputs {
-        walk_list!(visitor, visit_attribute, argument.attrs.iter());
-        visitor.visit_pat(&argument.pat);
-        visitor.visit_ty(&argument.ty);
-    }
-    visitor.visit_fn_ret_ty(&function_declaration.output)
+    function_declaration.inputs.iter().for_each(|arg| visitor.visit_arg(arg));
```
Comment (oli-obk) [inline src/libsyntax/visit.rs]: I don't like using for_each on containers too much. It should mainly be used for long iterator chains

File: src/test/ui/lint/lint-unused-mut-variables.rs [hunk 4]
```diff
@@ -4,9 +4,18 @@
 #![allow(unused_variables)]
 #![allow(dead_code)]
 #![deny(unused_mut)]
+#![feature(param_attrs)]
 
+//fn baz(#[allow(unused_mut)] mut b: i32) {}
```
Comment (oli-obk) [inline src/test/ui/lint/lint-unused-mut-variables.rs]: Why are these commented out?

File: src/test/ui/lint/lint-unused-variables.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+// Exercise the unused_mut attribute in some positive and negative cases
```
Comment (oli-obk) [inline src/test/ui/lint/lint-unused-variables.rs]: This test is about unused variables, not unused mut

File: src/librustc/hir/intravisit.rs [hunk 6]
```diff
@@ -396,10 +400,7 @@ pub fn walk_mod<'v, V: Visitor<'v>>(visitor: &mut V, module: &'v Mod, mod_hir_id
 }
 
 pub fn walk_body<'v, V: Visitor<'v>>(visitor: &mut V, body: &'v Body) {
-    for argument in &body.arguments {
-        visitor.visit_id(argument.hir_id);
-        visitor.visit_pat(&argument.pat);
-    }
+    body.arguments.iter().for_each(|arg| visitor.visit_arg(arg));
```
Comment (oli-obk) [inline src/librustc/hir/intravisit.rs]: Keep the for loop

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/test/ui/lint/lint-unused-mut-variables.rs (hunk 4 above)
Comment (c410-f3r) [reply src/test/ui/lint/lint-unused-mut-variables.rs]: I forgot to include a `FIXME` because it is currently failing.
Still trying to figure out how to make this lint work

File: src/libsyntax/visit.rs (hunk 3 above)
Comment (c410-f3r) [reply src/libsyntax/visit.rs]: Nit: `for_each` because it cleaner to have one line instead of three and this is not a "breakable" loop

File: src/test/ui/lint/lint-unused-variables.rs (hunk 5 above)
Comment (c410-f3r) [reply src/test/ui/lint/lint-unused-variables.rs]: Oopsss :S

File: src/test/ui/suggestions/as-ref.stderr (hunk 1 above)
Comment (c410-f3r) [reply src/test/ui/suggestions/as-ref.stderr]: I am investigating it

**Comment by Centril** [comment]
I believe Oliver won't be able to review this anytime soon so... r? @petrochenkov

File: src/librustc/hir/intravisit.rs [hunk 7]
```diff
@@ -396,9 +400,8 @@ pub fn walk_mod<'v, V: Visitor<'v>>(visitor: &mut V, module: &'v Mod, mod_hir_id
 }
 
 pub fn walk_body<'v, V: Visitor<'v>>(visitor: &mut V, body: &'v Body) {
-    for argument in &body.arguments {
-        visitor.visit_id(argument.hir_id);
-        visitor.visit_pat(&argument.pat);
+    for arg in &body.arguments {
```
Comment (petrochenkov) [inline src/librustc/hir/intravisit.rs]:
```suggestion
    walk_list!(visitor, visit_arg, &body.arguments);
```
visitor.rs and intravisit.rs have this macro for visiting anything iterable like lists or options, it's used almost everywhere.

File: src/librustc/hir/intravisit.rs [hunk 8]
```diff
@@ -210,6 +210,10 @@ pub trait Visitor<'v> : Sized {
         }
     }
 
+    fn visit_arg(&mut self, arg: &'v Arg) {
```
Comment (petrochenkov) [inline src/librustc/hir/intravisit.rs]: Well, the proper naming would be `param`, but that would introduce some discord since the existing code around the new code already uses `arg`.
Renaming everything could be a refactoring material for a separate PR.

File: src/librustc/middle/region.rs [hunk 9]
```diff
@@ -1429,6 +1429,9 @@ impl<'tcx> Visitor<'tcx> for RegionResolutionVisitor<'tcx> {
         self.terminating_scopes = outer_ts;
     }
 
+    fn visit_arg(&mut self, arg: &'tcx Arg) {
+        resolve_pat(self, &arg.pat);
+    }
```
Comment (petrochenkov) [inline src/librustc/middle/region.rs]: This should be unnecessary?
The default visitor already visits patterns in the arguments.

File: src/test/ui/rfc-2565-param-attrs/param-attrs-allowed.rs [hunk 10]
```diff
@@ -5,83 +5,83 @@
 
 extern "C" {
     fn ffi(
-        #[allow(C)] a: i32,
+        #[allow(unused_mut)] a: i32,
```
Comment (petrochenkov) [inline src/test/ui/rfc-2565-param-attrs/param-attrs-allowed.rs]: Lints from the `unused` group are `allow`ed by default in UI tests, so `allow` annotations in them don't test the normal behavior of `allow` unless there's a `#[warn(...)]` at the top of the file.

Comment (petrochenkov) [reply src/test/ui/rfc-2565-param-attrs/param-attrs-allowed.rs]: Same in `param-attrs-feature-gate.rs`.

File: src/librustc/hir/intravisit.rs (hunk 8 above)
Comment (Centril) [reply src/librustc/hir/intravisit.rs]: Filed https://github.com/rust-lang/rust/issues/62426 for a potential separate PR.

File: src/test/ui/suggestions/as-ref.stderr (hunk 1 above)
Comment (petrochenkov) [reply src/test/ui/suggestions/as-ref.stderr]: Perhaps some parent-child relations that the suggestion relies on were broken by the new nodes in HIR?

**Comment by petrochenkov** [comment]
The existing code looks good, but I don't know what exactly needs to be done to fix the regressions and remaining tests.

File: src/librustc/middle/region.rs (hunk 9 above)
Comment (c410-f3r) [reply src/librustc/middle/region.rs]: It probably came in some of my attempts to make `unused_mut` work

**Comment by c410-f3r** [comment]
Thanks for the review, @petrochenkov.

> The existing code looks good, but I don't know what exactly needs to be done to fix the regressions and remaining tests.

The biggest problem is the `unused_mut` lint. Someway, somehow, the `DataflowResultsConsumer` implementation for `MirBorrowckCtxt` must be modified so the `used_mut` attribute can be filled with the function attributes and the lint won't triggered [in this line code](https://github.com/rust-lang/rust/blob/04a3dd8a872633ca1e4c217d11f741cc35cb19a5/src/librustc_mir/borrow_check/mod.rs#L321).

**Comment by matthewjasper** [comment]
So, to fix the `unused_mut` lint requires giving the parameters the correct source scopes in MIR construction. The steps to do this are:

* Change the third field of this struct to `Option<(&'tcx hir::Arg)>` (
  https://github.com/rust-lang/rust/blob/dbeed58adee2ef046b46b252980f86672f9bfc4c/src/librustc_mir/build/mod.rs#L514

* Enter a new source scope inside this `if let`:
https://github.com/rust-lang/rust/blob/dbeed58adee2ef046b46b252980f86672f9bfc4c/src/librustc_mir/build/mod.rs#L824-L863
    * The relevant code for this is here, you can extract it into a new method first. `current_hir_id` is the `HirId` of the `Arg` that you added, span of the pattern for the `Span`.
      https://github.com/rust-lang/rust/blob/dbeed58adee2ef046b46b252980f86672f9bfc4c/src/librustc_mir/build/scope.rs#L426-L441

    * At the end of the `if let` you should reset the source scope to its previous value.

* Add `self.local_decls[local].source_info.scope = self.source_scope;` around here:
  https://github.com/rust-lang/rust/blob/dbeed58adee2ef046b46b252980f86672f9bfc4c/src/librustc_mir/build/mod.rs#L837

**Comment by c410-f3r** [comment]
@matthewjasper Thank you for joining this PR.
I may have missed something because the lint is not working yet (shame on me).
Changing the `HirId` of `Arg` to the `HirId` of the argument pattern didn't work either

**Comment by matthewjasper** [comment]
You also need to add a case for `Node::Arg` to this method: https://github.com/rust-lang/rust/blob/78ca1bda3522b14bc0336bc01dd1d49fdba2cda7/src/librustc/hir/map/mod.rs#L932-L953

File: src/test/ui/lint/lint-unused-mut-variables.rs [hunk 11]
```diff
@@ -1,12 +1,40 @@
...
-#![allow(unused_assignments)]
-#![allow(unused_variables)]
-#![allow(dead_code)]
 #![deny(unused_mut)]
+#![feature(param_attrs)]
+
+fn baz(
+    mut a: i32,
+    //~^ ERROR: variable does not need to be mutable
+    #[allow(unused_mut)] mut b: i32
```
Comment (matthewjasper) [inline src/test/ui/lint/lint-unused-mut-variables.rs]: Could you also add a tests for a more complex pattern, like `(mut x, y): (i32, i32)`?

File: src/librustc_mir/build/mod.rs [hunk 12]
```diff
@@ -872,6 +875,23 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+        &mut self,
+        arg_hir_id: hir::HirId,
+        pattern_span: Span
+    ) {
+        let tcx = self.hir.tcx();
+        let root = tcx.maybe_lint_level_root_bounded(
+            arg_hir_id,
+            self.hir.root_lint_level
+        );
+        self.source_scope = self.new_source_scope(
```
Comment (matthewjasper) [inline src/librustc_mir/build/mod.rs]: A new source scope only needs to be created if `root != self.hir.root_lint_scope`

**Comment by c410-f3r** [comment]
Thanks @matthewjasper

**Comment by c410-f3r** [comment]
`unused_variables` and `unused_mut` are now working and the regressions were fixed

File: src/test/ui/lint/lint-unused-mut-variables.rs [hunk 13]
```diff
@@ -1,12 +1,43 @@
...
+        #[allow(unused_mut)] (mut c, d): (i32, i32)
+    ) {}
+}
 
+trait RefTrait {
+    fn baz(
+        &self,
+        mut a: i32,
+        //~^ ERROR: variable does not need to be mutable
+        #[allow(unused_mut)] mut b: i32
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply src/test/ui/lint/lint-unused-mut-variables.rs]: A tuple type that was a correctly destructed pattern is located at line 30

[the remark this answers is not in the document]
Comment (c410-f3r) [reply src/test/ui/lint/lint-unused-mut-variables.rs]: Not in traits: `patterns aren't allowed in methods without bodies [E0642]`

File: src/librustc/hir/lowering.rs [hunk 14]
```diff
@@ -3281,6 +3282,7 @@ impl<'a> LoweringContext<'a> {
                 // original argument, but with a different pattern id.
                 let (new_argument_pat, new_argument_id) = this.pat_ident(desugared_span, ident);
                 let new_argument = hir::Arg {
+                    attrs: argument.attrs,
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply src/librustc/hir/lowering.rs]: I did not quite understand why it needs to be placed below. Nevertheless, async tests were included

[the remark this answers is not in the document]
Comment (c410-f3r) [reply src/librustc/hir/lowering.rs]: Not sure why it needs to be placed below. Nevertheless, async tests were included

**Comment by c410-f3r** [comment]
@matthewjasper Please ignore my deleted comments in your recent review

**Comment by c410-f3r** [comment]
ping @matthewjasper

File: src/librustc_typeck/check/_match.rs [hunk 15]
```diff
@@ -551,21 +551,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            let binding_parent = tcx.hir().get(binding_parent_id);
+            debug!("inner {:?} pat {:?} parent {:?}", inner, pat, binding_parent);
+            match binding_parent {
+                hir::Node::Arg(hir::Arg { pat, .. }) => {
                     if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
-                        // FIXME: turn into structured suggestion, will need a span that also
-                        // includes the the arg's type.
-                        err.help(&format!("did you mean `{}: &{}`?", snippet, expected));
+                        err.span_suggestion(
+                            pat.span,
```
Comment (matthewjasper) [inline src/librustc_typeck/check/_match.rs]: Sorry, I thought that `Arg` had it's own `Span`. This needs the span of the whole `Arg`.

File: src/librustc/hir/lowering.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (matthewjasper) [reply src/librustc/hir/lowering.rs]: Sorry, I meant a test for with lint attributes.

File: src/librustc_typeck/check/_match.rs (hunk 15 above)
Comment (c410-f3r) [reply src/librustc_typeck/check/_match.rs]: I added `span` for the whole argument and now 61 failed lldb errors are shown in my console. This doesn't seem trivial, can we revert this change and let another PR take care of this FIXME?

Comment (c410-f3r) [reply src/librustc_typeck/check/_match.rs]: Done

File: src/librustc/hir/lowering.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (c410-f3r) [reply src/librustc/hir/lowering.rs]: Parameters of async functions are now working properly

File: src/librustc/hir/map/mod.rs [hunk 16]
```diff
@@ -995,6 +997,7 @@ impl<'hir> Map<'hir> {
     pub fn span(&self, hir_id: HirId) -> Span {
         self.read(hir_id); // reveals span from node
         match self.find_entry(hir_id).map(|entry| entry.node) {
+            Some(Node::Arg(arg)) => arg.pat.span,
```
Comment (matthewjasper) [inline src/librustc/hir/map/mod.rs]: This should be `arg.span` now

Comment (c410-f3r) [reply src/librustc/hir/map/mod.rs]: In this new commit, is the `print_arg` function too simple? Maybe it isn't necessary?
